transformer_engine/common/util/cuda_runtime.cpp

```diff
@@ -0,0 +1,132 @@
+/*************************************************************************
+ * Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ *
+ * See LICENSE for license information.
+ ************************************************************************/
+
+#include <filesystem>
+#include <mutex>  // NOLINT(*)
+
+#include "../common.h"
+#include "../util/cuda_runtime.h"
+#include "../util/system.h"
+
+namespace transformer_engine {
+
+namespace cuda {
+
+int num_devices() {
+  auto query_num_devices = [] () -> int {
+    int count;
+    NVTE_CHECK_CUDA(cudaGetDeviceCount(&count));
+    return count;
+  };
+  static int num_devices_ = query_num_devices();
+  return num_devices_;
+}
+
+int current_device() {
+  int device_id;
+  NVTE_CHECK_CUDA(cudaGetDevice(&device_id));
```

Comment:
In CUDA 12 this could error out (with cudaErrorNoDevice) if the CUDA context is not yet initialized. We should handle that I guess?